// @cloudbase/database/dist/esm/realtime/snapshot.js
var Snapshot = (function () {
    function Snapshot(options) {
        var id = options.id, docChanges = options.docChanges, docs = options.docs, msgType = options.msgType, type = options.type;
        var cachedDocChanges;
        var cachedDocs;
        Object.defineProperties(this, {
            id: {
                get: function () { return id; },
                enumerable: true
            },
            docChanges: {
                get: function () {
                    if (!cachedDocChanges) {
                        cachedDocChanges = JSON.parse(JSON.stringify(docChanges));
                    }
                    return cachedDocChanges;
                },
                enumerable: true
            },
            docs: {
                get: function () {
                    if (!cachedDocs) {
                        cachedDocs = JSON.parse(JSON.stringify(docs));
                    }
                    return cachedDocs;
                },
                enumerable: true
            },
            msgType: {
                get: function () { return msgType; },
                enumerable: true
            },
            type: {
                get: function () { return type; },
                enumerable: true
            }
        });
    }
    return Snapshot;
}());
export { Snapshot };
